chainntnfs/neutrinonotify/queue.go

```diff
@@ -0,0 +1,79 @@
+package neutrinonotify
+
+import (
+	"container/list"
+)
+
+// chanCapacity is the capacity of the out channel to reduce overhead.
+const chanCapacity = 10
+
+// ConcurrentQueue is a concurrent-safe FIFO queue with unbounded capacity.
+// Clients interact with the queue by pushing items into the in channel and
+// popping items from the out channel. There is a goroutine that manages moving
+// items from the in channel to the out channel in the correct order that must
+// be started by calling Start().
+type ConcurrentQueue struct {
+	chanIn  chan interface{}
+	chanOut chan interface{}
+	quit    chan struct{}
+	items   *list.List
+}
+
+// NewConcurrentQueue constructs a ConcurrentQueue.
+func NewConcurrentQueue() *ConcurrentQueue {
+	return &ConcurrentQueue{
+		chanIn:  make(chan interface{}),
+		chanOut: make(chan interface{}, chanCapacity),
+		quit:    make(chan struct{}),
+		items:   list.New(),
+	}
+}
+
+// ChanIn returns a channel that can be used to push new items into the queue.
+func (cq *ConcurrentQueue) ChanIn() chan<- interface{} {
+	return cq.chanIn
+}
+
+// ChanOut returns a channel that can be used to pop items from the queue.
+func (cq *ConcurrentQueue) ChanOut() <-chan interface{} {
+	return cq.chanOut
+}
+
+// Start begins a goroutine that manages moving items from the in channel to the
+// out channel. This must be called before using the queue.
+func (cq *ConcurrentQueue) Start() {
+	go func() {
+		for {
+			nextElement := cq.items.Front()
+			if nextElement == nil {
```

Comment:
IMHO this case tree is a bit hard to read... would it be possible to handle the "optimistic" and "non-optimistic" case separately, to make it more clear? Maybe something like
```
for {
item <- chanIn:
  case chanOut <- item:
  default : 
    pushBack(item)
    break Loop
}

for len(items) > 0 {
 case chanOut <- items.Pop():
  case item <- chanIn:
    items.PushBack(item)
}
```
I feel the condition variable used in the above linked PR is easier to understand?

Reply:
I'm not really clear how your code example works as it's missing some selects and stuff. I added comments to this method and renamed `items` to `overflow`. Is it more clear now?

I prefer this approach to this condition variable as 1) it's simpler in that it just uses native channels instead of fancier synchronization stuff and 2) I believe it's more efficient since there are no mutexs and in the optimistic case (which should happen almost all of the time), it is just a channel receive and a channel send.